imap/jmap_core.c

```diff
@@ -399,6 +415,171 @@ static int getblob_cb(const conv_guidrec_t* rec, void* vrock)
+                size_t len = buf_len(&ctx.blob);
+                if (range.offset <= len) {
+                    base += range.offset;
+                    len -= range.offset;
+                }
+                else {
+                    r = -1;
+                }
+                if (range.length <= len) {
+                    if (range.length) len = range.length;
+                }
+                else {
+                    r = -1;
+                }
+                if (!r && len) {
+                    json_t *item = json_object();
+                    json_array_append_new(get.list, item);
+                    json_object_set_new(item, "id", json_string(blob_id));
+
+                    // the various data types
+                    int want_text = 0;
+                    int want_base64 = 0;
+                    if (jmap_wantprop(get.props, "data:asText")) {
+                        want_text = 1;
+                    }
+                    if (jmap_wantprop(get.props, "data:asBase64")) {
+                        want_base64 = 1;
+                    }
+                    if (jmap_wantprop(get.props, "data")) {
+                        struct char_counts guess_counts = charset_count_validutf8(base, len);
+                        if (!guess_counts.invalid && !guess_counts.replacement)
+                            want_text = 1;
+                        else
+                            want_base64 = 1;
+                    }
+                    if (want_text) {
+                        json_object_set_new(item, "data:asText", json_stringn(base, len));
+                    }
+                    if (want_base64) {
+                        size_t len64 = 0;
+                        charset_encode_mimebody(NULL, len, NULL, &len64, NULL, 0 /* no wrap */);
+                        char *encbuf = xzmalloc(len64+1);
+                        charset_encode_mimebody(base, len, encbuf, &len64, NULL, 0 /* no wrap */);
+                        json_object_set_new(item, "data:asBase64", json_stringn(encbuf, len64));
+                        free(encbuf);
+                    }
+                    if (jmap_wantprop(get.props, "data:asHex")) {
+                        char *encbuf = xzmalloc(len*2);
+                        bin_to_lchex(base, len, encbuf);
+                        json_object_set_new(item, "data:asHex", json_stringn(encbuf, len*2));
+                        free(encbuf);
+                    }
+
+                    if (jmap_wantprop(get.props, "size")) {
+                        // always the size of the full blob
+                        json_object_set_new(item, "size", json_integer(buf_len(&ctx.blob)));
+                    }
+                }
+                else {
+                    json_array_append_new(get.not_found, json_string(blob_id));
+                }
+            }
+            jmap_getblob_ctx_fini(&ctx);
+        }
+    }
+
+    /* Reply */
+    jmap_ok(req, jmap_get_reply(&get));
```

Comment:
See my comment above, we might want to `jmap_error` here in case of error.

Reply:
I think if we parsed OK, then unless there's a really bad error, we just want to return notFound, but... yeah, maybe we could bail and error if there's a real system error or something.